File: frontend/src/services/api/client-reports.ts
/**
 * Client reports (v5.261.0) — the findings-first Quarto deliverable, its
 * history and its addenda.  Backend: app/api/v1/endpoints/client_reports.py.
 * A draft's preview is an ordinary report job (poll with getReportJob,
 * download with downloadReportJob); an issued report's files download here.
 */
import { api, p } from './client';
import type { ReportJob } from '../api';

export type ClientReportKind = 'full' | 'addendum';
export type ClientReportStatus = 'draft' | 'issued' | 'superseded';
/** `qmd` is the Quarto source as a zip (report.qmd, data.json, filters, screenshots). */
export type ClientReportFormat = 'html' | 'docx' | 'pdf' | 'qmd';

export interface ReportTester {
  user_id?: number | null;
  name: string;
  role?: string | null;
  email?: string | null;
}

export interface ReportRecipient {
  name: string;
  email?: string | null;
}

export interface EngagementSettings {
  client_name: string | null;
  classification: string | null;
  engagement_type: string | null;
  testers: ReportTester[];
  distribution: ReportRecipient[];
  system_description: string | null;
  /** v5.263.0 — the other target lists ("if applicable"; Markdown). */
  applications?: string | null;
  thick_clients?: string | null;
  other_targets?: string | null;
}

export interface ReportProfile extends EngagementSettings {
  template: string | null;
  updated_at?: string | null;
  /** No team is saved: `testers` is the project's analysts and admins. */
  testers_from_project?: boolean;
}

/** An image the template itself expects (logo, cover art) — declared in its
 *  template.json, never finding evidence.  `path` is relative to
 *  `report-templates/<template name>/` on the server. */
export interface ReportTemplateAsset {
  id: string;
  path: string;
  label: string;
  description: string;
  /** Extra guidance from the template author (e.g. a Word header lives in reference.docx). */
  note: string;
  /** A missing required image blocks preview, issue and render. */
  required: boolean;
  formats: ClientReportFormat[];
  /** When installed, used in place of this template file (e.g. `reference.docx`). */
  replaces?: string | null;
  /** The file is installed where the manifest says. */
  present: boolean;
}

export interface ReportTemplate {
  name: string;
  title: string;
  description: string;
  formats: ClientReportFormat[];
  assets?: ReportTemplateAsset[];
}

export interface ReportFile {
  format: ClientReportFormat;
  filename: string;
  media_type: string;
  size_bytes: number;
  sha256: string;
  created_at: string | null;
}

export interface ReportRef {
  id: number;
  number: number | null;
  title: string;
  status: ClientReportStatus;
  issued_at: string | null;
}

export interface ReportSummary {
  counts?: Record<'critical' | 'high' | 'medium' | 'low' | 'info' | 'total', number>;
  findings_shown?: number;
  under_investigation?: number;
  missing_text?: Array<{ id: number; ref: string; title: string; missing: string[] }>;
  /** v5.263.0 — report details still empty (printed as a highlighted TODO). */
  missing_details?: string[];
  images?: number;
  images_skipped?: number;
  delta?: { new_findings: number; findings_with_new_endpoints: number; withdrawn: number } | null;
  /** Set when the summary could not be built (e.g. an addendum lost its baseline). */
  error?: string;
}

export interface ClientReport {
  id: number;
  project_id: number;
  kind: ClientReportKind;
  status: ClientReportStatus;
  title: string;
  number: number | null;
  template: string;
  baseline: ReportRef | null;
  revision_of: ReportRef | null;
  superseded_by: ReportRef | null;
  settings: EngagementSettings;
  executive_summary: string | null;
  template_fingerprint: string | null;
  quarto_version: string | null;
  render_status: 'pending' | 'done' | 'failed' | null;
  render_error: string | null;
  files: ReportFile[];
  created_by_name: string | null;
  issued_by_name: string | null;
  created_at: string | null;
  updated_at: string | null;
  issued_at: string | null;
  summary: ReportSummary | null;
  can_edit: boolean;
  can_issue: boolean;
}

export interface ClientReportList {
  items: ClientReport[];
  latest_issued_id: number | null;
  can_create: boolean;
  can_issue: boolean;
}

export interface ClientReportUpdate {
  title?: string;
  template?: string;
  baseline_report_id?: number;
  executive_summary?: string | null;
  settings?: EngagementSettings;
}

const base = () => `${p()}/client-reports`;

export const listClientReports = async (): Promise<ClientReportList> =>
  (await api.get<ClientReportList>(base())).data;

export const getClientReport = async (id: number): Promise<ClientReport> =>
  (await api.get<ClientReport>(`${base()}/${id}`)).data;

export const createClientReport = async (body: {
  kind: ClientReportKind; title?: string; template?: string; baseline_report_id?: number;
}): Promise<ClientReport> => (await api.post<ClientReport>(base(), body)).data;

export const updateClientReport = async (id: number, body: ClientReportUpdate): Promise<ClientReport> =>
  (await api.patch<ClientReport>(`${base()}/${id}`, body)).data;

export const deleteClientReport = async (id: number): Promise<void> => {
  await api.delete(`${base()}/${id}`);
};

export const previewClientReport = async (id: number, format: ClientReportFormat): Promise<ReportJob> =>
  (await api.post<ReportJob>(`${base()}/${id}/preview`, { format })).data;

export const issueClientReport = async (id: number): Promise<ClientReport> =>
  (await api.post<ClientReport>(`${base()}/${id}/issue`)).data;

export const rerenderClientReport = async (id: number): Promise<ClientReport> =>
  (await api.post<ClientReport>(`${base()}/${id}/render`)).data;

export const reviseClientReport = async (id: number): Promise<ClientReport> =>
  (await api.post<ClientReport>(`${base()}/${id}/revise`)).data;

export const listReportTemplates = async (): Promise<ReportTemplate[]> =>
  (await api.get<ReportTemplate[]>(`${base()}/templates`)).data;

/** The project's analysts and admins as an assessment team (name, role, email). */
export const getProjectReportTeam = async (): Promise<ReportTester[]> =>
  (await api.get<ReportTester[]>(`${base()}/team`)).data;

export const getReportProfile = async (): Promise<ReportProfile> =>
  (await api.get<ReportProfile>(`${base()}/profile`)).data;

export const saveReportProfile = async (body: Omit<ReportProfile, 'updated_at'>): Promise<ReportProfile> =>
  (await api.put<ReportProfile>(`${base()}/profile`, body)).data;

/** Save an issued report's file (authenticated blob → browser download). */
export const downloadClientReportFile = async (id: number, file: ReportFile): Promise<void> => {
  const response = await api.get(`${base()}/${id}/files/${file.format}`, { responseType: 'blob' });
  const url = window.URL.createObjectURL(new Blob([response.data], { type: file.media_type }));
  const a = document.createElement('a');
  a.href = url;
  a.download = file.filename;
  document.body.appendChild(a);
  a.click();
  window.URL.revokeObjectURL(url);
  document.body.removeChild(a);
};

File: backend/app/schemas/client_reports.py
"""Schemas for client reports (v2.380.0)."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


class Tester(BaseModel):
    user_id: Optional[int] = None
    name: str = Field(..., min_length=1, max_length=200)
    role: Optional[str] = Field(None, max_length=200)
    email: Optional[str] = Field(None, max_length=254)


class Recipient(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    email: Optional[str] = Field(None, max_length=254)


class EngagementSettings(BaseModel):
    """Engagement details — the profile's defaults, or one report's copy."""
    client_name: Optional[str] = Field(None, max_length=255)
    classification: Optional[str] = Field(None, max_length=100)
    engagement_type: Optional[str] = Field(None, max_length=100)
    testers: List[Tester] = Field(default_factory=list, max_length=50)
    distribution: List[Recipient] = Field(default_factory=list, max_length=100)
    system_description: Optional[str] = Field(None, max_length=32768)
    # v2.382.0 — the template's other target lists ("if applicable"; Markdown).
    # Networks and domains come from the project's scope.
    applications: Optional[str] = Field(None, max_length=32768)
    thick_clients: Optional[str] = Field(None, max_length=32768)
    other_targets: Optional[str] = Field(None, max_length=32768)

    @field_validator(
        "client_name", "classification", "engagement_type", "system_description",
        "applications", "thick_clients", "other_targets",
    )
    @classmethod
    def _blank_is_none(cls, v):
        if v is None:
            return None
        v = v.strip()
        return v or None


class ReportProfileBody(EngagementSettings):
    template: Optional[str] = Field(None, max_length=100)


class ReportProfileOut(ReportProfileBody):
    updated_at: Optional[datetime] = None
    # v2.382.0 — no team is saved, so `testers` is the project's analysts and
    # admins: what a new draft will list.
    testers_from_project: bool = False


class ReportTemplateAssetOut(BaseModel):
    """An image the template itself expects (logo, cover art), from its
    template.json — never finding evidence.  ``path`` is relative to
    ``report-templates/<name>/``."""
    id: str
    path: str
    label: str
    description: str = ""
    note: str = ""
    required: bool = False
    formats: List[str] = []
    # When installed, used in place of this template file (e.g. reference.docx).
    replaces: Optional[str] = None
    present: bool = False


class ReportTemplateOut(BaseModel):
    name: str
    title: str
    description: str
    formats: List[str]
    assets: List[ReportTemplateAssetOut] = []


class ReportFileOut(BaseModel):
    format: str
    filename: str
    media_type: str
    size_bytes: int
    sha256: str
    created_at: Optional[datetime] = None


class ReportRef(BaseModel):
    id: int
    number: Optional[int] = None
    title: str
    status: str
    issued_at: Optional[datetime] = None


class ReportOut(BaseModel):
    id: int
    project_id: int
    kind: str
    status: str
    title: str
    number: Optional[int] = None
    template: str
    baseline: Optional[ReportRef] = None
    revision_of: Optional[ReportRef] = None
    superseded_by: Optional[ReportRef] = None
    settings: EngagementSettings
    executive_summary: Optional[str] = None
    template_fingerprint: Optional[str] = None
    quarto_version: Optional[str] = None
    render_status: Optional[str] = None
    render_error: Optional[str] = None
    files: List[ReportFileOut] = []
    created_by_name: Optional[str] = None
    issued_by_name: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    issued_at: Optional[datetime] = None
    # What the report contains (live for a draft, frozen for an issued one).
    summary: Optional[Dict[str, Any]] = None
    # What the CALLER may do.
    can_edit: bool = False
    can_issue: bool = False


class ReportListOut(BaseModel):
    items: List[ReportOut]
    # The report an addendum is compared against by default.
    latest_issued_id: Optional[int] = None
    can_create: bool = False
    can_issue: bool = False


class ReportCreate(BaseModel):
    kind: Literal["full", "addendum"] = "full"
    title: Optional[str] = Field(None, max_length=255)
    template: Optional[str] = Field(None, max_length=100)
    # Addendum only; defaults to the latest issued report.
    baseline_report_id: Optional[int] = None


class ReportUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    template: Optional[str] = Field(None, max_length=100)
    baseline_report_id: Optional[int] = None
    executive_summary: Optional[str] = Field(None, max_length=65536)
    settings: Optional[EngagementSettings] = None


class PreviewRequest(BaseModel):
    format: Literal["html", "docx", "qmd"] = "html"
